cpp/src/lists/drop_list_duplicates.cu

```diff
@@ -0,0 +1,291 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/copy.hpp>
+#include <cudf/detail/gather.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/lists/drop_list_duplicates.hpp>
+#include <cudf/lists/sorting.hpp>
+#include <cudf/table/row_operators.cuh>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/binary_search.h>
+#include <thrust/transform.h>
+
+namespace cudf {
+namespace lists {
+namespace detail {
+using offset_type = lists_column_view::offset_type;
+/**
+ * @brief Copy list entries and entry list offsets ignoring duplicates
+ *
+ * Given an array of all entries flattened from a list column and an array that maps each entry to
+ * the offset of the list containing that entry, those entries and list offsets are copied into
+ * new arrays such that the duplicated entries within each list will be ignored.
+ *
+ * @param all_lists_entries    The input array containing all list entries
+ * @param entries_list_offsets A map from list entries to their corresponding list offsets
+ * @param nulls_equal          Flag to specify whether null entries should be considered equal
+ * @param stream               CUDA stream used for device memory operations and kernel launches
+ * @param mr                   Device resource used to allocate memory
+ *
+ * @return A pair of columns, the first one contains unique list entries and the second one
+ * contains their corresponding list offsets
+ */
+template <bool has_nulls>
+std::vector<std::unique_ptr<column>> get_unique_entries_and_list_offsets(
+  column_view const& all_lists_entries,
+  column_view const& entries_list_offsets,
+  null_equality nulls_equal,
+  rmm::cuda_stream_view stream,
+  rmm::mr::device_memory_resource* mr)
+{
+  // Create an intermediate table, since the comparator only work on tables
+  auto const device_input_table =
+    cudf::table_device_view::create(table_view{{all_lists_entries}}, stream);
+  auto const comp = row_equality_comparator<has_nulls>(
+    *device_input_table, *device_input_table, nulls_equal == null_equality::EQUAL);
+
+  auto const num_entries = all_lists_entries.size();
+  // Allocate memory to store the indices of the unique entries
+  auto const unique_indices = cudf::make_numeric_column(
+    entries_list_offsets.type(), num_entries, mask_state::UNALLOCATED, stream);
+  auto const unique_indices_begin = unique_indices->mutable_view().begin<offset_type>();
+
+  auto const copy_end = thrust::unique_copy(
```

Comment:
Would `thrust::unique_by_key` here work instead? The keys would be `entries_list_offsets`.

Reply:
I  thought about that too. However, I don't know how to make thrust handling (copying/removing) cudf rows, thus I had to create unique_indices and use `cudf::detail::gather` to collect cudf rows from those indices.

Reply:
Right, you'd still have to operate on the indices, but you'd use a `unique_by_key` instead of `unique_copy` with a predicate that checks `list_offsets[i] == list_offsets[j]`.

Reply:
If doing so, I will have to fill the array `unique_indices` a sequence `[0, 1, 2, ...]` first (with `thrust::fill_n`), then call `thrust::unique_by_key` on that array. Do you have a justification to use those two thrust APIs instead of one? Otherwise, let's keep using `thrust::unique_copy`.

Reply:
Dave just proposed to use `unique_by_key_copy` to achive our desired behavior. I'm looking into it :)

Reply:
Yeah, sorry, I forgot that `unique_by_key` expects to be in-place. `unique_by_key_copy` is what you want so that you can just use a counting iterator for the input indices.

Reply:
I decided to keep using `unique_copy`, as `unique_by_key_copy` does exactly the same thing but it requires more parameters.

Reply:
I'm pretty sure the `unique_by_key` is going to be more efficient. Not only that, it much more clearly expresses your intent.

Reply:
I just tried to apply `unique_by_key_copy` without success. It requires 2 output iterators. We only want to output to one array. As its documentation says it is just "a generalization of unique_copy", then let's keep using `unique_copy`. https://thrust.github.io/doc/group__stream__compaction_ga3ce9e9746ed1a166cefa9dc08c1f875a.html#ga3ce9e9746ed1a166cefa9dc08c1f875a

Reply:
You can use a thrust discard iterator to ignore the other output (`thrust::make_discard_iterator()`)

Reply:
What was the resolution here?

Reply:
I missed it, too. Was the resolution to use `make_discard_iterator()`?